### PR TITLE
Exp replay bug

### PR DESCRIPTION
**Pull request by NikolaosBouas**
Fixes the bug where states were appended with the previous reward to the replay buffer.

### CODE REVIEW

File: src/expreplay.py [hunk 1]
```diff
@@ -132,6 +146,16 @@ def _assign(self, pos, exp):
             self.reward[i, pos] = exp[2][i]
             self.isOver[i, pos] = exp[3][i]
 
+    def _assign_state(self, pos, obs):
+        for i in range(self.agents):
+            self.state[i, pos] = obs[i]
```
Comment (gml16) [inline src/expreplay.py]: If both `self.state` and `obs` are numpy arrays, could we avoid the for loop and make the assignment faster with something like `self.state[:, pos] = obs[:]`? Similarly for the method `_assign_effect`

File: src/DQNModel.py [hunk 2]
```diff
@@ -335,10 +335,11 @@ def _calculate_loss(self, transitions, discount_factor):
         curr_state = torch.tensor(transitions[0])
         next_state = torch.tensor(transitions[3])
         terminal = torch.tensor(transitions[4]).type(torch.int)
-
+
```
Comment (gml16) [inline src/DQNModel.py]: Nitpicking but could you remove the changes in this file if it's just white spaces?

File: src/run_evaluation.py [hunk 3]
```diff
@@ -21,11 +21,15 @@
...
     parser.add_argument(
-        '--files', type=argparse.FileType('r'), nargs='+',
+        '--model_files', type=argparse.FileType('r'), nargs='+',
         help="Filepath to the models that must be evaluated")
-    # parser.add_argument(
-    #     '--fixed_spawn', nargs='*',  type=float,
-    #     help='Starting position of the agents during rollout. Randomised if not specified.',)
+
+    parser.add_argument(
+        '--files', type=argparse.FileType('r'), nargs='+',
```
Comment (gml16) [inline src/run_evaluation.py:28]: Thanks for removing my hardcoded values and adding an argument :)

File: src/trainer.py [hunk 4]
```diff
@@ -83,18 +83,21 @@ def train(self):
...
                 acc_steps += 1
+                # Step the agent once, and get the transition tuple
+                index = self.buffer.append_obs(obs)
                 acts, q_values = self.get_next_actions(
                     self.buffer.recent_state())
-                # Step the agent once, and get the transition tuple
-                obs, reward, terminal, info = self.env.step(
+                next_obs, reward, terminal, info = self.env.step(
                     np.copy(acts), q_values, terminal)
+                self.buffer.append_effect((index, obs, acts, reward, terminal))
```
Comment (gml16) [inline src/trainer.py:98]: Thank you for fixing this bug

File: src/trainer.py [hunk 5]
```diff
@@ -169,6 +175,8 @@ def append_epoch_board(self, epoch_dists, eps=0, losses=[],
         epoch_dists = np.array(epoch_dists)
         if name == "train":
             lr = self.dqn.scheduler.state_dict()["_last_lr"]
+            if isinstance(lr, list):
```
Comment (gml16) [inline src/trainer.py:178]: In what case can the learning rate be a list here? Shouldn't it always be a float?

**Review by gml16 — CHANGES_REQUESTED** [review]
Thanks a lot Nikos for the PR :) I've added a few comments

File: src/trainer.py (hunk 5 above)
Comment (NikolaosBouas) [reply src/trainer.py:178]: It returned a list and threw an error when I tried to print it in the next line, if I didn't take the first element of the list. It seems like the _last_lr is a list if you look at the last line of the step function [here](https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html#StepLR).

**Review by gml16 — APPROVED** [review]

File: src/trainer.py (hunk 5 above)
Comment (gml16) [reply src/trainer.py:178]: Oh yes that makes sense, thanks
